features/telemetry/protocols/bmp/adapter/src/main/java/org/opennms/netmgt/telemetry/protocols/bmp/adapter/BmpMessagePersister.java

```diff
@@ -540,6 +600,120 @@ private BmpGlobalIpRib buildGlobalIpRib(PrefixByAS prefixByAS) {
 
     }
 
+    private BmpAsnInfo fetchAndBuildAsnInfo(Long asn) {
+        Optional<AsnInfo> asnInfoOptional = BmpWhoIsClient.getAsnInfo(asn);
+        if (asnInfoOptional.isPresent()) {
+            BmpAsnInfo bmpAsnInfo = new BmpAsnInfo();
+            AsnInfo asnInfo = asnInfoOptional.get();
+            bmpAsnInfo.setAsn(asnInfo.getAsn());
+            bmpAsnInfo.setOrgId(asnInfo.getOrgId());
+            bmpAsnInfo.setAsName(asnInfo.getAsName());
+            bmpAsnInfo.setOrgName(asnInfo.getOrgName());
+            bmpAsnInfo.setAddress(asnInfo.getAddress());
+            bmpAsnInfo.setCity(asnInfo.getCity());
+            bmpAsnInfo.setStateProv(asnInfo.getStateProv());
+            bmpAsnInfo.setPostalCode(asnInfo.getPostalCode());
+            bmpAsnInfo.setCountry(asnInfo.getCountry());
+            bmpAsnInfo.setSource(asnInfo.getSource());
+            bmpAsnInfo.setRawOutput(asnInfo.getRawOutput());
+            bmpAsnInfo.setRemarks(asnInfo.getRemarks());
+            bmpAsnInfo.setLastUpdated(Date.from(Instant.now()));
+            return bmpAsnInfo;
+        }
+        return null;
+    }
+
+    private BmpRouteInfo fetchAndBuildRouteInfo(String prefix) {
+        Optional<RouteInfo> routeInfoOptional = BmpWhoIsClient.getRouteInfo(prefix);
+        if (routeInfoOptional.isPresent() && routeInfoOptional.get().getPrefix() != null) {
+            RouteInfo routeInfo = routeInfoOptional.get();
+            Integer prefixLen = routeInfo.getPrefixLen();
+            Long originAs = routeInfo.getOriginAs();
+            BmpRouteInfo bmpRouteInfo = bmpRouteInfoDao.findByPrefix(routeInfo.getPrefix(), prefixLen, originAs);
+            if (bmpRouteInfo == null) {
+                bmpRouteInfo = new BmpRouteInfo();
+                bmpRouteInfo.setPrefix(routeInfo.getPrefix());
+                bmpRouteInfo.setPrefixLen(routeInfo.getPrefixLen());
+                bmpRouteInfo.setDescr(routeInfo.getDescription());
+                bmpRouteInfo.setOriginAs(routeInfo.getOriginAs());
+                bmpRouteInfo.setSource(routeInfo.getSource());
+            }
+            bmpRouteInfo.setLastUpdated(Date.from(Instant.now()));
+            return bmpRouteInfo;
+        }
+        return null;
+    }
+
+    private List<BmpAsnPathAnalysis> buildBmpAsnPath(String asnPath) {
+
+        List<BmpAsnPathAnalysis> bmpAsnPathAnalyses = new ArrayList<>();
+        String[] asnArray = asnPath.split(" ");
+
+        Long leftAsn = 0L;
+        Long rightAsn = 0L;
+        Long asn = 0L;
+        for (int i = 0; i < asnArray.length; i++) {
```

Comment:
The loop can be simplified by iterating from `i = 1` to `i = asnArray.length - `. The  you can use `i-1` and `i+1` for left and right ASN inside the loop.

In addition, the Origin AS handling can be done out of loop at the end.